### ISSUE TITLE
GET plugin API returns weak ETag validator

### ISSUE THREAD

**Issue by ankitsri11** [gocd/gocd#6278]
##### Issue Type

- Feature enhancement

##### Summary

Currently, GET plugin API returns "W/" in the ETag identifier whereas GET specific environment API returns ETag identifier without "W/".

https://api.gocd.org/current/#get-plugin-settings

For example:

```
curl -kv 'https://localhost:8154/go/api/admin/plugin_settings/com.thoughtworks.gocd.analytics' \
        -H 'Authorization: bearer XXXXXXXXXXXXX' \
      -H 'Accept:application/vnd.go.cd.v1+json' \
      -i

*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8154 (#0)
.
.
.
< ETag: W/"fd6d6c7c005834d62981a0c5505144ea"
ETag: W/"fd6d6c7c005834d62981a0c5505144ea"
.
.
.

```

**Comment by arvindsv** [comment]
@ankitsri11 Whether it is a weak etag or not _should_ not matter. It should be an implementation detail.  Technically, providing whatever comes back, when making an update should work. Can you try using those different ETags and seeing what happens?

**Comment by ankitsri11** [comment]
@arvindsv: Here is an example which I tried updating plugin setting using PUT request with and without **W/**:

- [GET request](https://api.gocd.org/current/#get-plugin-settings) to get the plugin settings

```
» curl -kv 'http://localhost:8153/go/api/admin/plugin_settings/com.thoughtworks.gocd.analytics' \                                                                                            
      -H 'Accept:application/vnd.go.cd.v1+json'

*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8153 (#0)
> GET /go/api/admin/plugin_settings/com.thoughtworks.gocd.analytics HTTP/1.1
> Host: localhost:8153
> User-Agent: curl/7.54.0
> Accept:application/vnd.go.cd.v1+json
>
< HTTP/1.1 200 OK
< Date: Mon, 13 May 2019 20:56:11 GMT
< X-XSS-Protection: 1; mode=block
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-UA-Compatible: chrome=1
< Set-Cookie: JSESSIONID=node01k9xr0l9npccrxpojeek4vwci3.node0;Path=/go;Expires=Mon, 27-May-2019 20:56:11 GMT;Max-Age=1209600;HttpOnly
< Expires: Thu, 01 Jan 1970 00:00:00 GMT
< ETag: W/"4d993b31d7d189c5c5c7bde4672b7a74"
.
.
.
.
```

- Update plugin settings (hostname update from "localhost" to "127.0.0.1") using the [PUT request](https://api.gocd.org/current/#update-plugin-settings).
  
  * using **W/** in the request

 ```
» curl 'http://localhost:8153/go/api/admin/plugin_settings/com.thoughtworks.gocd.analytics' \                                                                                                
-H 'Accept: application/vnd.go.cd.v1+json' \
-H 'Content-Type: application/json' \
-H 'If-Match: "W/4d993b31d7d189c5c5c7bde4672b7a74"' \
-X PUT -d '{
  "plugin_id": "com.thoughtworks.gocd.analytics",
  "configuration": [
  {
      "key": "client_cert"
    },
    {
      "key": "use_ssl"
    },
    {
      "key": "client_pkcs8_key"
    },
    {
      "key": "max_connections_active"
    },
    {
      "key": "root_cert"
    },
    {
      "key": "client_key"
    },
    {
      "key": "license",
      "value": "***you_license_key*******"
    },
    {
      "key": "password",
      "encrypted_value": "AES:******"
    },
    {
      "key": "periodic_cleanup_time"
    },
    {
      "key": "port",
      "value": "5432"
    },
    {
      "key": "host",
      "value": "127.0.0.1"
    },
    {
      "key": "name",
      "value": "plugin"
    },
    {
      "key": "max_connections_idle"
    },
    {
      "key": "connection"
    },
    {
      "key": "ssl_mode"
    },
    {
      "key": "max_connection_wait_time"
    },
    {
      "key": "username",
      "value": "postgres"
    }
  ]
}'
{
  "message": "Someone has modified the configuration for Plugin Settings 'com.thoughtworks.gocd.analytics'. Please update your copy of the config with the changes."
}
```

  * Without W/ in the request, which resulted in success and updated the plugin settings.

```
curl 'http://localhost:8153/go/api/admin/plugin_settings/com.thoughtworks.gocd.analytics' \
-H 'Accept: application/vnd.go.cd.v1+json' \
-H 'Content-Type: application/json' \
-H 'If-Match: "4d993b31d7d189c5c5c7bde4672b7a74"' \
-X PUT -d '{
  "plugin_id": "com.thoughtworks.gocd.analytics",
  "configuration": [
 {
      "key": "client_cert"
    },
    {
      "key": "use_ssl"
    },
    {
      "key": "client_pkcs8_key"
    },
    {
      "key": "max_connections_active"
    },
    {
      "key": "root_cert"
    },
    {
      "key": "client_key"
    },
    {
      "key": "license",
      "value": "***you_license_key*******"
    },
    {
      "key": "password",
      "encrypted_value": "AES:******"
    },
    {
      "key": "periodic_cleanup_time"
    },
    {
      "key": "port",
      "value": "5432"
    },
    {
      "key": "host",
      "value": "127.0.0.1"
    },
    {
      "key": "name",
      "value": "plugin"
    },
    {
      "key": "max_connections_idle"
    },
    {
      "key": "connection"
    },
    {
      "key": "ssl_mode"
    },
    {
      "key": "max_connection_wait_time"
    },
    {
      "key": "username",
      "value": "postgres"
    }
  ]
}'
```

**Comment by arvindsv** [comment]
Thanks, @ankitsri11. I think the expectation should be that it works equally well with the `W/` too.

/cc @ketan

**Comment by GaneshSPatil** [comment]
@arvindsv @ankitsri11 @ketan @maheshp -- 

To answer [this](https://github.com/gocd/gocd/issues/6278#issuecomment-491993987) comment.
The behavior of the Plugin Settings API for ETag is appropriate.

According to [ETag RFC](https://tools.ietf.org/html/rfc7232#section-2.3), An entity-tag consists of an opaque quoted string, possibly prefixed by a weakness indicator.
Example:
* ETag: "xyzzy"
* ETag: W/"xyzzy"

In both examples, the ETag value is "xyzzy". Whereas, in the second example, `W/` represents its a weak ETag.

---

**NOTE:** **The plugin settings API has started returning Weak ETag from GoCD version `v18.10.0`.**

### Why did Plugin Settings API start returning weak ETag after GoCD version `v18.10.0`?

As part of GoCD release `v18.10.0`, we've upgraded a bunch of libraries to add support for Java 10 and 11.  We've **upgraded rails from rails-4 to rails-5** which has caused this behavior change.

##### Looking at [the release notes of `rails 5`](https://guides.rubyonrails.org/5_0_release_notes.html), it is mentioned that:
1. Rails generates weak ETags by default. [Pull Request](https://github.com/rails/rails/pull/17573)
2. Introduce Response#strong_etag= and #weak_etag= and analogous options for fresh_when and stale?. [Pull Request](https://github.com/rails/rails/pull/24387)

This is a breaking change, which has changed the behavior of `stale?(etag: "etag")` to return weak ETag in `rails5`.

---

### Why did Rails 5 switches from strong Etag to weak Etag?

Rails does not perform strong validation of ETags as implied by [strong ETags spec](https://tools.ietf.org/html/rfc7232#section-2.1). Rails just check whether the incoming ETag from the request headers matches with the ETag of the generated response. It does not do byte by byte comparison of the response.

**This behavior was true even before Rails 5. So this change is more of a course correction.** 

---

### How do we fix this issue?

Changing `stale?(etag: "etag")` to `stale?(strong_etag: "etag")` will fix the issue.

### Is any other API apart from Plugin Settings affected?

Yes. As this bug was introduced due to rails upgrade, all the rails based GoCD APIs will have this bug.
Example:
1. Package Repositories
2. Pluggable SCMs
3. Get All Resources (Internal API used on Agents SPA)
4. Authorization

**Comment by arvindsv** [comment]
#6309 should fix this and make sure that all generated etags are strong.

**Comment by rajiesh** [comment]
#6309 fixes etag for all rails APIs except the Index calls like Get all Auth config API. The fix needs to be relooked, moving this to 19.5

cc @GaneshSPatil

**Comment by rajiesh** [comment]
Verified on 19.5.0 (9272-496bf8b95e603c1f3980ae59042bc559eecbbbc0)
